D-Freezer/freezer.md

```diff
@@ -0,0 +1,186 @@
+
+```
+const struct sched_class fair_sched_class = {
+	.next			= &idle_sched_class,
+	.enqueue_task		= enqueue_task_fair,
+	.dequeue_task		= dequeue_task_fair,
+	.yield_task		= yield_task_fair,
+	.yield_to_task		= yield_to_task_fair,
+
+	.check_preempt_curr	= check_preempt_wakeup,
+
+	.pick_next_task		= pick_next_task_fair,
+	.put_prev_task		= put_prev_task_fair,
+
+#ifdef CONFIG_SMP
+	.select_task_rq		= select_task_rq_fair,
+	.migrate_task_rq	= migrate_task_rq_fair,
+
+	.rq_online		= rq_online_fair,
+	.rq_offline		= rq_offline_fair,
+
+	.task_dead		= task_dead_fair,
+	.set_cpus_allowed	= set_cpus_allowed_common,
+#endif
+
+	.set_curr_task          = set_curr_task_fair,
+	.task_tick		= task_tick_fair,
+	.task_fork		= task_fork_fair,
+
+	.prio_changed		= prio_changed_fair,
+	.switched_from		= switched_from_fair,
+	.switched_to		= switched_to_fair,
+
+	.get_rr_interval	= get_rr_interval_fair,
+
+	.update_curr		= update_curr_fair,
+
+#ifdef CONFIG_FAIR_GROUP_SCHED
+	.task_change_group	= task_change_group_fair,
+#endif
+};
+```
+
+Note, the dot notation is a C99 feature that allows you to set specific fields of the struct by name in an initializer. Also, not every function needs to be implemented. You will need to figure out what is and is not necessary.  
+
+As you can see, CFS initializes the `struct sched_class` function pointers to the CFS implementation. Two things of note here. First, the convention is to name the struct `<class_name>_sched_class`, so CFS names it `fair_sched_class`. The second thing of note is to name a particular class's functions as `<function_name>_<class_name>`. For example, the CFS implementation of `enqueue_task` as `enqueue_task_fair`. Now, every time the Kernel needs to call function, it can call can simply call `p->sched_class-><function()>`. Here, `p` is of the type `task_struct *`, `sched_class` is a pointer within the `task_struct` pointing to an instanece of `struct sched_class`, and the `<function()>` points the spesific implementaion of the the function to be called(). 
+
+One final thing, you may have noticed the first member of `struct sched_class` of `const struct sched_class *next`. That is because all the scheduling classes are linked together on a singly linked list like this. 
+<div align='center'>
+    <img src='./allclass.png'/><br/>
+</div>
+
+The first class on the list is of higher priority than the second. In other words, `sched_class_dl` has a higher priority than `sched_class_rt`. Now, every time a new process needs to be scheduled, the schedular can simply go through the class list and check if there is a process of that class that needs to run. If there isn't, it will move to the next class.  
```

Comment:
nice. if you think its worth it, maybe we can briefly say how this happens

schedule() -> pick_next_task() -> and then for_each_class (see here: https://elixir.bootlin.com/linux/v4.19.50/source/kernel/sched/core.c#L3305)

Reply:
I just feel that this section doesnt wrap up on the narrative you created: "Say we have a CFS task about to return from main(), we need to call the CFS `dequeue_task()` to remove it from the CFS queue. How can we ensure that the Kernel will call the CFS implementation of `dequeue_task()`? "

Saying something about how schedule() is eventually called somewhere in the OS and briefly running through some of the functions like I mentioned above might be a good way to to end this recitation note.

Reply:
After adding something about the idle_sched_class (see my comment above), you could also add a quite sentence here about how idle_sched_class is the lowest priority schedule which explains why it only kicks in when nothing else is available